binderhub/build.py

```diff
@@ -630,3 +694,89 @@ def stream_logs(self):
                 }
             ),
         )
+
+
+class Build(KubernetesBuildExecutor):
+    """DEPRECATED: Use KubernetesBuildExecutor and configure with Traitlets
+
+    Represents a build of a git repository into a docker image.
+
+    This ultimately maps to a single pod on a kubernetes cluster. Many
+    different build objects can point to this single pod and perform
+    operations on the pod. The code in this class needs to be careful and take
+    this into account.
+
+    For example, operations a Build object tries might not succeed because
+    another Build object pointing to the same pod might have done something
+    else. This should be handled gracefully, and the build object should
+    reflect the state of the pod as quickly as possible.
+
+    ``name``
+        The ``name`` should be unique and immutable since it is used to
+        sync to the pod. The ``name`` should be unique for a
+        ``(repo_url, ref)`` tuple, and the same tuple should correspond
+        to the same ``name``. This allows use of the locking provided by k8s
+        API instead of having to invent our own locking code.
+
+    """
+
+    """
+    For backwards compatibility: BinderHub previously assumed a single cleaner for all builds
+    """
+    _cleaners = {}
+
+    def __init__(
+        self,
+        q,
+        api,
+        name,
+        *,
+        namespace,
+        repo_url,
+        ref,
+        build_image,
+        docker_host,
+        image_name,
+        git_credentials=None,
+        push_secret=None,
+        memory_limit=0,
+        memory_request=0,
+        node_selector=None,
+        appendix="",
+        log_tail_lines=100,
+        sticky_builds=False,
+    ):
+        warnings.warn(
+            "Class Build is deprecated, use KubernetesBuildExecutor and configure with Traitlets",
+            DeprecationWarning,
+        )
+
+        super().__init__()
+
+        self.q = q
+        self.api = api
+        self.repo_url = repo_url
+        self.ref = ref
+        self.name = name
+        self.namespace = namespace
+        self.image_name = image_name
+        self.push_secret = push_secret
+        self.build_image = build_image
+        self.memory_limit = memory_limit
+        self.memory_request = memory_request
+        self.docker_host = docker_host
+        self.node_selector = node_selector
+        self.appendix = appendix
+        self.log_tail_lines = log_tail_lines
+        self.git_credentials = git_credentials
+        self.sticky_builds = sticky_builds
+
+    @classmethod
+    def cleanup_builds(cls, kube, namespace, max_age):
```

Comment:
BinderHub hard-codes the cleanup command here:
https://github.com/jupyterhub/binderhub/blob/0fa35a06e7f7d054f8bb8dd2cba158ed7f45ac68/binderhub/app.py#L940-L944

Though since the `Build` class is hardcoded maybe it's fine to get rid of the `cleanup_builds()` method since even if it's overridden it'll never be used?

Reply:
If we switch the default to the new traitlets class (https://github.com/jupyterhub/binderhub/pull/1521) we won't need to worry about this.